infra/modules/worklytics-connectors-msft-365/main.tf

```diff
@@ -22,36 +22,50 @@ module "worklytics_connector_specs" {
 }
 
 locals {
-  todos_to_populate = { for k, v in module.worklytics_connector_specs.enabled_msft_365_connectors : k => v if try(v.external_token_todo != null, false) && var.todos_as_local_files }
+  provision_entraid_apps = var.msft_connector_app_object_id == null
+  connectors_needing_apps = local.provision_entraid_apps ? module.worklytics_connector_specs.enabled_msft_365_connectors : {}
 }
 
 data "azuread_client_config" "current" {
+  count = local.provision_entraid_apps ? 1 : 0
 
 }
 
 data "azuread_users" "owners" {
+  count = local.provision_entraid_apps && length(var.msft_owners_email) > 0 ? 1 : 0
+
   user_principal_names = var.msft_owners_email
 }
 
+
 module "msft_connection" {
-  for_each = module.worklytics_connector_specs.enabled_msft_365_connectors
+  for_each = local.connectors_needing_apps
 
   source = "../../modules/azuread-connection"
 
   display_name                      = "Psoxy Connector - ${each.value.display_name}${local.environment_id_display_name_qualifier}"
   tenant_id                         = var.msft_tenant_id
   required_app_roles                = each.value.required_app_roles
   required_oauth2_permission_scopes = each.value.required_oauth2_permission_scopes
-  owners = toset(concat(data.azuread_users.owners.object_ids, [
-    data.azuread_client_config.current.object_id
+  owners = toset(concat(data.azuread_users.owners[0].object_ids, [
+    data.azuread_client_config.current[0].object_id
   ]))
 }
 
+# if an existing app object id is provided, use it as a shared app for ALL MSFT connectors
+# (requires that it has the superset of permissions required by all connectors)
+data "azuread_application" "existing_connector_app" {
+  count = var.msft_connector_app_object_id == null ? 0 : 1
+
+  object_id = var.msft_connector_app_object_id
+}
+
+
 # grant required permissions to connectors via Azure AD
 # (requires terraform configuration being applied by an Azure User with privileges to do this; it
 #  usually requires a 'Global Administrator' for your tenant)
 module "msft_365_grants" {
-  for_each = module.worklytics_connector_specs.enabled_msft_365_connectors
+  for_each = local.connectors_needing_apps
```

Comment:
Another approach, maybe we could split the todos in two scenarios: the case of the provided app id OR the regular case where we create all the apps.

In case of providing the app id, having a single TODO with the link to grant the permissions and maybe with a `set` we can put the complete list of permissions required from spec and the external todo. In the regular case, we'll follow existing approach.

Reply:
yeah, working on this; needs a distinct TODO file instead of this, instructing to make the grant.